RFC: API Change: PWM: add support for inverted PWM signals

henrikbrixandersen opened the pull request:
Add support for requesting an inverted PWM pulse (active-low) when setting up the period and pulse width of a PWM pin. This is useful when driving external, active-low circuitry with a PWM signal.

You can do a "poor man" inverted PWM pulse just by "inverting" the pulse width, but without proper API support for inverted PWM signals it is not possible to do edge-aligned PWM with either a mix of active-low/active-high or any number of active-low PWM signals (which is often needed for motor control).

This PR updates all in-tree PWM controller drivers to match the new API signature. The NXP Kinetis FlexTimer (FTM) PWM controller driver has furthermore been updated to support PWM flags set through the device tree (as an example implementation).

The following in-tree samples have been updated to support PWM flags:
- samples/basic/blink_led
- samples/basic/fade_led
- samples/basic/rgb_led

Also, the PWM shell commands has been updated to allow specifying PWM flags. All other in-tree PWM consumers are updated to pass a flags value of 0.

The goal is to enable specification of PWM signal polarity through the device tree for supporting PWM controllers (e.g. using `pwms = <&pwm0 1 PWM_POLARITY_INVERTED>;` for PWM controller 0 channel/pin 1). Please see the included changes for the FTM PWM driver for examples.

Closes #21384

henrikbrixandersen committed:
drivers: pwm: add support for inverted PWM signals

Add support for requesting an inverted PWM pulse (active-low) when
setting up the period and pulse width of a PWM pin. This is useful
when driving external, active-low circuitry (e.g. an LED) with a PWM
signal.

All in-tree PWM drivers is updated to match the new API signature, but
no driver support for inverted PWM signals is added yet.

All in-tree PWM consumers are updated to pass a flags value of 0
(0 meaning default, which is normal PWM polarity).

Fixes #21384.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen commented:
The following PWM drivers can easily be updated to support this new API:
- `pwm_nrfx.c` (already supports inverted PWM signals through custom DTS properties)
- `pwm_mcux_ftm.c` (implemented through custom DTS properties in #20263)

I expect other PWM peripherals already support setting PWM signal polarity. If a given PWM peripheral does not have hardware support for inverted signals, it can easily be handled within the driver itself, if desired.

carlescufi commented:
As per API meeting, there are no major concerns with the change but the PWM API is currently marked as stable, which would force us to deprecate the current functions. This needs discussion at the TSC level to decide whether moving an API from stable to unstable is warranted (coupled with other examples in the tree).

henrikbrixandersen commented on include/dt-bindings/pwm/pwm.h:
```diff
@@ -0,0 +1,27 @@
+#define PWM_POL_INV	(1 << 0)
+
+/** @cond INTERNAL_HIDDEN */
+#define PWM_POL_MASK	0x1
```
The naming of these defines was chosen to mimic the currently used GPIO defines in `include/dt-bindings/gpio/gpio.h`. As pointed out on the API meeting, the naming scheme should be changed to match the flags naming scheme of the `topic-gpio` branch (e.g. `PWM_POLARITY_NORMAL` instead of the abbreviated `PWM_POL_NORMAL`).

henrikbrixandersen commented on include/drivers/pwm.h:
```diff
@@ -34,7 +34,8 @@ extern "C" {
 typedef int (*pwm_pin_set_t)(struct device *dev, u32_t pwm,
-			     u32_t period_cycles, u32_t pulse_cycles);
+			     u32_t period_cycles, u32_t pulse_cycles,
+			     int flags);
```
Another comment from the API meeting: flags ought to be unsigned (e.g. `u32_t`). The implementation here uses `int` since that is what is used for the flags in the Zephyr GPIO API (both on master and on topic-gpio).

Should we consider changing this for GPIO as well?

henrikbrixandersen replied:
According to @mnkp this change is already planned for GPIO. I have changed the suggested API for PWM flags to use `unsigned int`.

henrikbrixandersen commented:
Input to the next API meeting: According to our [API Lifecycle documentation](https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#stable), non-backwards compatible changes to a stable API (The PWM API is currently marked as stable) can be done but must be mentioned in the release notes. The documentation also says that "Backwards-compatibility [for a stable API] will be maintained if reasonable".

henrikbrixandersen committed:
drivers: pwm: mcux_ftm: add support for inverted polarity pwm signals

Add support for specifying the PWM signal polarity through flags to
the NXP Kinetis FlexTimer (FTM) PWM driver.

Prior to this change the FTM PWM driver always produced inverted
polarity (active-low) PWM signals.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
samples: blink_led: add support for PWM flags

Pass PWM flags set in the device tree, if any, to the
pwm_pin_set_usec() functions.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
samples: rgb_led: add support for PWM flags

Pass PWM flags set in the device tree, if any, to the
pwm_pin_set_usec() functions.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
drivers: pwm: shell: add PWM flags support

Add support for specifying flags in the PWM shell commands.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
dts: nxp: kinetis-ftm: add PWM flags cell

Add support for specifying PWM flags for the NXP Kinetis FlexTimer
(FTM) PWM driver through the device tree.

All in-tree clients of this PWM controller are active-low LEDs.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
samples: fade_led: add support for PWM flags

Pass PWM flags set in the device tree, if any, to the
pwm_pin_set_usec() function.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

carlescufi commented:
@henrikbrixandersen please follow [the new procedure](https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#introducing-incompatible-changes) for this change.

henrikbrixandersen commented:
> 
> 
> @henrikbrixandersen please follow [the new procedure](https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#introducing-incompatible-changes) for this change.

@carlescufi I have submitted an RFC issue (#21384). I did my best to follow the new instructions, but please let me know if I missed/misunderstood anything.

carlescufi commented:
> > @henrikbrixandersen please follow [the new procedure](https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#introducing-incompatible-changes) for this change.
> 
> @carlescufi I have submitted an RFC issue (#21384). I did my best to follow the new instructions, but please let me know if I missed/misunderstood anything.

Just went through it, looks great, thank you. Remember to email the mailing list.

carlescufi reviewed (changes requested):
Missing addition to the 2.2 release notes.
See https://github.com/zephyrproject-rtos/zephyr/pull/20375/commits/fba74500446f6be9af86ee1064f48bc21572aa26 for an example.

carlescufi commented:
@StefJar can you please review?

henrikbrixandersen committed:
doc: release notes: Update contents for 2.2

List the added PWM API flags parameter under changes to stable APIs
for the 2.2 release.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen commented:
> Missing addition to the 2.2 release notes.

@carlescufi Thanks, added.

StefJar commented:
> @StefJar can you please review?

Think this is fine so far as I can say from checking the files. It is not breaking the existing driver code. Let's see when the first drivers are getting "pwm inverted ready" ;)

dbkinder reviewed (approved):
doc changes LGTM, thanks.

henrikbrixandersen commented:
> Think this is fine so far as I can say from checking the files.

Thank you for reviewing.

> It is not breaking the existing driver code. Let's see when the first drivers are getting "pwm inverted ready" ;)

This PR already contains one driver conversion (NXP FTM PWM controller). I have the changes for the Nordic Semi nRF PWM driver ready, but I'd like to get #21071 and #21051 in first.

anangl reviewed (commented):
Looks good. One question, though. Is there a particular reason why `flags` is of the type `unsigned int`, not `u32_t`?

henrikbrixandersen commented:
> Looks good. One question, though. Is there a particular reason why `flags` is of the type `unsigned int`, not `u32_t`?

@anangl I used `unsigned int` since that is what is used for similar purposes in the `topic-gpio` branch.

anangl commented:
@henrikbrixandersen 
> I can see that the `topic-gpio` branch has changed to using a u32_t typedef for flags (https://github.com/zephyrproject-rtos/zephyr/blob/topic-gpio/include/drivers/gpio.h#L432). Should I do the same for the PWM flags for consistency?

I think it would be good. You could use `pwm_flags_t` and actually define it as `u8_t`, as it would be sufficient to hold all the currently available flags.

henrikbrixandersen commented:
@anangl
> I think it would be good. You could use `pwm_flags_t` and actually define it as `u8_t`, as it would be sufficient to hold all the currently available flags.

Implemented as you suggested. Please review.

anangl commented on samples/basic/rgb_led/src/main.c:
```diff
@@ -44,9 +59,9 @@
 
 static int write_pin(struct device *pwm_dev, u32_t pwm_pin,
-		     u32_t pulse_width)
+		     u32_t pulse_width, unsigned int flags)
```
`pwm_flags_t flags`

henrikbrixandersen replied:
@anangl Fixed. Thank you for reviewing.

anangl reviewed (approved):
Spotted one more occurrence of `flags` to be updated. Otherwise LGTM.

mnkp reviewed (approved):
The two comments are minor, just something to consider.
